pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,457 @@
+
+	// VerificationData build verification data which can be used for signing or verification.
+	VerificationData() ([]byte, error)
+
+	// LookupHeader makes look up of particular header.
+	LookupHeader(name string) (interface{}, bool)
+
+	// LookupStringHeader makes look up of particular header with string value.
+	LookupStringHeader(name string) string
+
+	// Signature returns a signature if defined.
+	Signature() []byte
+
+	// CompactSerialize makes compact serialization of token.
+	CompactSerialize(detached bool) (string, error)
+}
+
+type jsonWebToken struct {
+	headers map[string]interface{}
+
+	payload map[string]interface{}
+
+	signature []byte
+}
+
+// CompactSerialize implements compact serialization of token.
+func (j *jsonWebToken) CompactSerialize(detached bool) (string, error) {
+	headers, err := toBase64(j.headers)
+	if err != nil {
+		return "", fmt.Errorf("base64 headers: %w", err)
+	}
+
+	payload := ""
+	if !detached {
+		payload, err = toBase64(j.payload)
+		if err != nil {
+			return "", fmt.Errorf("base64 payload: %w", err)
+		}
+	}
+
+	signature := base64.RawURLEncoding.EncodeToString(j.signature)
+
+	return fmt.Sprintf("%s.%s.%s", headers, payload, signature), nil
+}
+
+// Claims fills input c with JWT claims taken from payload of a token.
+func (j *jsonWebToken) DecodeClaims(c interface{}) error {
+	pBytes, err := json.Marshal(j.payload)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(pBytes, c)
+}
+
+// VerificationData build verification data which can be used for signing or verification.
+func (j *jsonWebToken) VerificationData() ([]byte, error) {
+	return jwsVerificationData(j.headers, j.payload)
+}
+
+// Header makes look up of particular header.
+func (j *jsonWebToken) LookupHeader(name string) (interface{}, bool) {
+	v, ok := j.headers[name]
+
+	return v, ok
+}
+
+// HeaderString makes look up of particular header with string value.
+func (j *jsonWebToken) LookupStringHeader(name string) string {
+	if headerValue, ok := j.headers[name]; ok {
+		if headerStrValue, ok := headerValue.(string); ok {
+			return headerStrValue
+		}
+	}
+
+	return ""
+}
+
+// Signature returns a copy of JWT signature if defined.
+func (j *jsonWebToken) Signature() []byte {
+	if j.signature == nil {
+		return nil
+	}
+
+	return append(j.signature[:0:0], j.signature...)
+}
+
+// Signer defines JWT Signer interface. It makes signing of data and provides custom JWT headers relevant to the signer.
+type Signer interface {
+	// Sign signs.
+	Sign(data []byte) ([]byte, error)
+
+	// Headers provides JWT headers. "alg" header must be provided (see https://tools.ietf.org/html/rfc7515#section-4.1)
+	Headers() map[string]interface{}
+}
+
+// Builder construct JSON Web Token.
+type Builder interface {
+	// Signed creates JWS.
+	Signed(signer Signer) (JSONWebToken, error)
+
+	// Unsecured creates unsecured JWT.
+	Unsecured(headers map[string]interface{}) (JSONWebToken, error)
+}
+
+// builder implements Builder.
+type builder struct {
```

Comment:
this struct isn't acting as a builder pattern. (nor is the associated interface.)